compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -2713,6 +2714,46 @@ private void checkConstantAssignment(BLangExpression varRef, AnalyzerData data)
         }
     }
 
+    private boolean checkIfFiniteTypeHasNullMembers(BType bType) {
+        if (bType.tag != TypeTags.FINITE) {
+            return false;
+        }
+        Set<BLangExpression> finiteMembers = ((BFiniteType) bType).getValueSpace();
+        for (BLangExpression finiteMemberType : finiteMembers) {
+            if (finiteMemberType.getBType().tag != TypeTags.NIL) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean checkIfUnionTypeHasNullMembers(BType bType) {
+        if (bType.tag != TypeTags.UNION) {
+            return false;
+        }
+        BUnionType unionType = (BUnionType) bType;
+        LinkedHashSet<BType> members = unionType.getMemberTypes();
+        if (members.size() >= 1) {
+            for (BType memberType : members) {
+                if (memberType.tag != TypeTags.NIL) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        LinkedHashSet<BType> originalMembers = unionType.getOriginalMemberTypes();
+        // Special cased to handle return types like never|never` as they are not correctly handled at the moment.
+        // https://github.com/ballerina-platform/ballerina-lang/issues/39797
+        if (members.size() == 0 && originalMembers.size() == 1) {
+            return originalMembers.iterator().next().tag == TypeTags.NEVER;
+        }
+        return false;
+    }
+
+    private boolean isSubTypeOfNil(BType bType) {
```

Comment:
We can refactor overall logic as this(need to verify whether it is okay to use streams in Java.)
```java
private boolean hasNullMembers(BType bType) {
    switch (bType.tag) {
        case TypeTags.UNION:
            BUnionType unionType = (BUnionType) bType;
            return unionType.getMemberTypes().stream()
                    .allMatch(memberType -> memberType.tag == TypeTags.NIL);
        case TypeTags.FINITE:
            Set<BLangExpression> finiteMembers = ((BFiniteType) bType).getValueSpace();
            return finiteMembers.stream()
                    .allMatch(finiteMemberType -> finiteMemberType.getBType().tag == TypeTags.NIL);
        default:
            return false;
    }
}
```

Reply:
As we use `isAssignable`, we can ignore this.